client/oidc.go

```diff
@@ -89,6 +99,52 @@ func VerifyPKToken(ctx context.Context, pkt *pktoken.PKToken, provider OpenIdPro
 	return nil
 }
 
+func DiscoverPublicKey(ctx context.Context, headersEnc []byte, issuer string) (crypto.PublicKey, error) {
+	headersJSON, err := util.Base64DecodeForJWT(headersEnc)
+	if err != nil {
+		return nil, err
+	}
+
+	headersMap := make(map[string]any)
+	err = json.Unmarshal(headersJSON, &headersMap)
+	if err != nil {
+		return nil, err
+	}
+
+	kidRaw, ok := headersMap["kid"]
+	if !ok {
+		return nil, fmt.Errorf("missing kid claim")
+	}
+
+	kid, ok := kidRaw.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected kid claim to be a string, got %T", kidRaw)
+	}
+
+	discConf, err := oidcclient.Discover(issuer, http.DefaultClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call OIDC discovery endpoint: %w", err)
+	}
+
+	jwks, err := jwk.Fetch(ctx, discConf.JwksURI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch to JWKS: %w", err)
+	}
+
+	key, ok := jwks.LookupKeyID(kid)
+	if !ok {
+		return nil, fmt.Errorf("key %q isn't in JWKS", kid)
+	}
+
+	pubKey := new(rsa.PublicKey)
```

Comment:
Before casting to a particular key type, say RSA, we should check that the algorithm of the returned JWK matches the key type we are casting to. This is similar to my comment above, but unless I am mistaken, I think we should do this check in both places.

Your PR does not introduce this issue as it exists in the current code, but it is a valuable check to have.

Reply:
I think this is already OK because the jwx library does this check itself. The key is parsed after being fetched and stored in an appropriate struct based on the `alg` field (see https://github.com/lestrrat-go/jwx/blob/262da1ab53fc42371eb8a13ae61044589b1a239b/jwk/jwk.go#L443-L467). Then `key.Raw(pubKey)` checks that `pubKey` is of the appropriate type for the type of `key` and errors otherwise.

Reply:
You are correct I misread what the code was doing here.

It might save a future dev some time if we create a custom error for the case where the key is not RS256. I've made a suggestion for this above. Not a requirement for this PR so feel free to reject the suggestion and resolve this comment.

Reply:
Yes, good idea. I'll tweak the suggestion slightly